plugins/janus_sip.c

```diff
@@ -2986,37 +3064,54 @@ static void *janus_sip_handler(void *data) {
 			}
 
 			session->account.registration_status = janus_sip_registration_status_registering;
-			if(!refresh && session->stack == NULL) {
+			if (!refresh) {
 				/* Start the thread first */
 				GError *error = NULL;
 				char tname[16];
 				g_snprintf(tname, sizeof(tname), "sip %s", session->account.username);
 				janus_refcount_increase(&session->ref);
 				g_thread_try_new(tname, janus_sip_sofia_thread, session, &error);
 				if(error != NULL) {
-					janus_refcount_decrease(&session->ref);
 					JANUS_LOG(LOG_ERR, "Got error %d (%s) trying to launch the SIP Sofia thread...\n",
 						error->code, error->message ? error->message : "??");
 					error_code = JANUS_SIP_ERROR_UNKNOWN_ERROR;
 					g_snprintf(error_cause, 512, "Got error %d (%s) trying to launch the SIP Sofia thread",
 						error->code, error->message ? error->message : "??");
 					g_error_free(error);
+					janus_refcount_decrease(&session->ref);
 					goto error;
 				}
 				long int timeout = 0;
-				while(session->stack == NULL || session->stack->s_nua == NULL) {
-					g_usleep(100000);
-					timeout += 100000;
-					if(timeout >= 2000000) {
+				while (timeout < 2000000) {
+					janus_mutex_lock(&session->stack->smutex);
+					if (session->stack->s_nua != NULL) {
+						janus_mutex_unlock(&session->stack->smutex);
 						break;
 					}
+					janus_mutex_unlock(&session->stack->smutex);
+					g_usleep(100000);
+					timeout += 100000;
```

Comment:
I don't understand the changes above, which seem unrelated to the main purpose of the patch.

Reply:
We create session->stack->s_nua in different thread. So, we have to use the mutex here. However, you are right - it's a part of different patch.